File: 03_Sliding_Window/01_Best_Time_to_Buy_and_Sell_Stock/0121-best-time-to-buy-and-sell-stock.py
"""
Problem: LeetCode 121 - Best Time to Buy and Sell Stock

Key Idea:
To find the maximum profit from a single buy and sell transaction in the input array 'prices', we can use a simple one-pass approach. We initialize two variables, 'min_price' to keep track of the minimum price seen so far, and 'max_profit' to store the maximum profit. We iterate through the 'prices' array, and for each price, we update the 'min_price' if we find a smaller price. We also calculate the potential profit if we sell at the current price and update 'max_profit' if the current profit is greater than the previous maximum profit.

Time Complexity:
The time complexity of this solution is O(n), where n is the number of elements in the input array 'prices'. We iterate through the array once to find the maximum profit, and at each step, we perform constant-time operations.

Space Complexity:
The space complexity is O(1) since we are not using any additional data structures that depend on the input size. We only use a constant amount of extra space for the variables to store the minimum price and maximum profit.
"""


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0

        min_price = float("inf")
        max_profit = 0

        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)

        return max_profit

File: 03_Sliding_Window/01_Best_Time_to_Buy_and_Sell_Stock/0121-best-time-to-buy-and-sell-stock.rs
/*
Problem: LeetCode 121 - Best Time to Buy and Sell Stock

Key Idea:
The key idea is to iterate through the prices while keeping track of the minimum price seen so far. Calculate the potential profit at each step by subtracting the minimum price from the current price, and update the maximum profit if a higher profit is found.

Approach:
1. Initialize variables `min_price` to represent the minimum price encountered so far and `max_profit` to represent the maximum profit.
2. Iterate through the prices:
   - For each price `p`, update `min_price` if `p` is smaller.
   - Calculate the potential profit by subtracting `min_price` from `p`.
   - Update `max_profit` with the maximum of its current value and the potential profit.
3. Return `max_profit` as the result.

Time Complexity:
O(n), where n is the number of elements in the input array. We perform a single pass through the array to find the maximum profit.

Space Complexity:
O(1), as we use only a constant amount of extra space for variables.
*/

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min_price = i32::max_value();
        let mut max_profit = 0;

        for price in prices {
            min_price = min_price.min(price);
            max_profit = max_profit.max(price - min_price);
        }

        max_profit
    }
}
